File: scripts/deduplicate-vocabulary.ts
import { readFile, writeFile } from 'node:fs/promises';

/**
 * Advanced vocabulary deduplication system
 * with semantic merging capabilities
 */
class VocabularyDeduplicator {
  private duplicateGroups: Map<string, any[]> = new Map();
  private mergeStrategies: Record<string, (items: any[]) => any> = {
    // Merge by semantic equivalence (same meaning)
    semantic: (items) => {
      // Keep the most complete version
      return items.reduce((best, current) => {
        const currentScore = this.calculateCompletenessScore(current);
        const bestScore = this.calculateCompletenessScore(best);
        return currentScore > bestScore ? current : best;
      });
    },

    // Merge aspect variations (perfective/imperfective verbs)
    aspect: (items) => {
      // Combine into single entry with aspect notes
      const baseItem = items[0];
      return {
        ...baseItem,
        metadata: {
          ...baseItem.metadata,
          notes: `${baseItem.metadata?.notes || ''}\n\nAspect variations: ${items.map(i => i.partOfSpeech).join(', ')}`
        },
        aspectVariations: items.map(item => ({
          id: item.id,
          partOfSpeech: item.partOfSpeech,
          examples: item.metadata?.examples || []
        }))
      };
    },

    // Merge POS variations (noun/adjective pairs)
    pos: (items) => {
      // Create separate entries with clear classification
      return items.map(item => ({
        ...item,
        categories: [...new Set([
          ...(item.categories || []),
          'grammar',
          item.partOfSpeech
        ])]
      }));
    }
  };

  private calculateCompletenessScore(item: any): number {
    let score = 0;
    if (item.metadata?.examples?.length) score += 10;
    if (item.metadata?.notes) score += 5;
    if (item.metadata?.examples?.every((e: any) => e.german && e.bulgarian)) score += 15;
    if (item.categories?.length && !item.categories.includes('uncategorized')) score += 5;
    return score;
  }

  private generateUniqueId(baseId: string, suffix: string): string {
    return `${baseId}-${suffix}-${Date.now().toString(36)}`;
  }

  async deduplicate(filePath: string): Promise<void> {
    const data = JSON.parse(await readFile(filePath, 'utf-8'));
    const originalCount = data.items.length;

    // Group duplicates by semantic key
    const semanticGroups = new Map<string, any[]>();
    data.items.forEach((item: any) => {
      const semanticKey = `${item.bulgarian.toLowerCase()}-${item.german.toLowerCase()}`;
      if (!semanticGroups.has(semanticKey)) {
        semanticGroups.set(semanticKey, []);
      }
      semanticGroups.get(semanticKey)?.push(item);
    });

    // Process each group
    const deduplicatedItems: any[] = [];
    const duplicateReport: any[] = [];

    semanticGroups.forEach((group, semanticKey) => {
      if (group.length === 1) {
        deduplicatedItems.push(group[0]);
        return;
      }

      // Determine merge strategy
      const strategy = this.determineMergeStrategy(group);
      const mergedItem = this.mergeStrategies[strategy](group);

      // Handle ID conflicts
      if (group.some(item => item.id === mergedItem.id)) {
        mergedItem.id = this.generateUniqueId(
          mergedItem.id.replace(/-.*$/, ''),
          strategy
        );
      }

      deduplicatedItems.push(mergedItem);
      duplicateReport.push({
        semanticKey,
        originalCount: group.length,
        strategy,
        mergedId: mergedItem.id,
        originalIds: group.map(item => item.id)
      });
    });

    // Update data
    const deduplicatedData = {
      ...data,
      items: deduplicatedItems,
      metadata: {
        ...data.metadata,
        deduplication: {
          timestamp: new Date().toISOString(),
          originalCount,
          deduplicatedCount: deduplicatedItems.length,
          reductionRate: `${Math.round(((originalCount - deduplicatedItems.length) / originalCount) * 100)}%`
        }
      }
    };

    // Save results
    await writeFile(filePath, JSON.stringify(deduplicatedData, null, 2));
    await writeFile(
      'data/vocabulary-deduplication-report.json',
      JSON.stringify({
        timestamp: new Date().toISOString(),
        duplicates: duplicateReport,
        summary: {
          originalCount,
          deduplicatedCount: deduplicatedItems.length,
          reductionRate: `${Math.round(((originalCount - deduplicatedItems.length) / originalCount) * 100)}%`
        }
      }, null, 2)
    );

    console.log(`✅ Deduplication complete`);
    console.log(`📊 Original items: ${originalCount}`);
    console.log(`📊 Deduplicated items: ${deduplicatedItems.length}`);
    console.log(`📈 Reduction rate: ${Math.round(((originalCount - deduplicatedItems.length) / originalCount) * 100)}%`);
    console.log(`📋 Report saved to data/vocabulary-deduplication-report.json`);
  }

  private determineMergeStrategy(group: any[]): string {
    // Check for aspect variations (verbs)
    const isAspectVariation = group.every(item =>
      item.partOfSpeech.includes('verb') &&
      (item.bulgarian.includes('(perfektiv)') || item.bulgarian.includes('(imperfektiv)'))
    );

    if (isAspectVariation) return 'aspect';

    // Check for POS variations
    const uniquePOS = new Set(group.map(item => item.partOfSpeech));
    if (uniquePOS.size > 1) return 'pos';

    // Default to semantic merge
    return 'semantic';
  }
}

// CLI Usage
const deduplicator = new VocabularyDeduplicator();
deduplicator.deduplicate(process.argv[2])
  .catch(console.error);

File: scripts/deduplicate-vocabulary.py
#!/usr/bin/env python3
"""Deduplicate vocabulary entries, keeping the most complete version of each."""

import json
from datetime import datetime
from pathlib import Path

def completeness_score(item):
    """Score an item by how complete it is - prefer items with more data"""
    score = 0
    if item.get('examples'): score += len(item['examples']) * 2
    if item.get('definition'): score += len(str(item['definition'])) // 50
    if item.get('culturalNotes'): score += 3
    if item.get('ipa'): score += 2
    if item.get('tags'): score += len(item['tags'])
    if item.get('relatedWords'): score += len(item['relatedWords'])
    if item.get('contextualUsage'): score += 2
    if item.get('etymology'): score += 2
    return score

def main():
    data_dir = Path(__file__).parent.parent / 'data'
    vocab_file = data_dir / 'unified-vocabulary.json'
    
    # Load vocabulary
    with open(vocab_file, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    vocab = data.get('vocabulary', data) if isinstance(data, dict) else data
    print(f"Total items before: {len(vocab)}")
    
    # Deduplicate, keeping the most complete version
    seen = {}
    for item in vocab:
        key = (item.get('german', '').strip().lower(), item.get('bulgarian', '').strip().lower())
        if key not in seen:
            seen[key] = item
        else:
            # Keep the more complete version
            if completeness_score(item) > completeness_score(seen[key]):
                seen[key] = item
    
    unique = list(seen.values())
    removed_count = len(vocab) - len(unique)
    print(f"Unique items after: {len(unique)}")
    print(f"Removed: {removed_count} duplicates")
    
    if removed_count == 0:
        print("No duplicates found. Exiting.")
        return
    
    # Create backup
    backup_name = data_dir / f"unified-vocabulary.backup-{datetime.now().strftime('%Y-%m-%d-%H%M')}.json"
    with open(backup_name, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    print(f"Backup saved to: {backup_name}")
    
    # Save deduplicated data
    output = {'vocabulary': unique} if isinstance(data, dict) and 'vocabulary' in data else unique
    with open(vocab_file, 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)
    
    print(f"✅ Deduplication complete! Saved {len(unique)} unique items.")

if __name__ == '__main__':
    main()
